### PR TITLE
Enhance async download project functionality

### PR DESCRIPTION
**Pull request by GoldenAnpu**

### CODE REVIEW

**Review by NikolaiPetukhov — APPROVED** [review]

File: supervisely/api/image_api.py [hunk 1]
```diff
@@ -4157,3 +4165,185 @@ async def download_bytes_many_async(
...
+                            f"Downloaded hash of image with ID:{img_id} does not match the expected hash: {downloaded_bytes_hash} != {hhash}"
+                        )
+            if progress_cb is not None and progress_cb_type == "number":
+                progress_cb(1)
+            elif progress_cb is not None and progress_cb_type == "size":
+                progress_cb(len(part.content))
+
+            yield img_id, part.content
+
+    async def get_list_page_generator_async(
```
Comment (almazgimaev) [inline supervisely/api/image_api.py:4255]: mb move it to `module_api.py` ?

File: supervisely/project/project.py [hunk 2]
```diff
@@ -4401,14 +4408,42 @@ async def _download_project_async(
...
-    progress_cb: Optional[Callable] = None,
+    progress_cb: Optional[Union[tqdm, Callable]] = None,
     save_image_meta: Optional[bool] = False,
     images_ids: Optional[List[int]] = None,
     resume_download: Optional[bool] = False,
+    **kwargs,
 ):
+    """
+    Download image project to the local directory asynchronously.
+    Uses qeueu and semaphore to control the number of parallel downloads.
```
Comment (almazgimaev) [inline supervisely/project/project.py:4419]: typo
qeueu > queue

File: supervisely/project/project.py [hunk 3]
```diff
@@ -4401,14 +4408,42 @@ async def _download_project_async(
...
+    Every image goes through size check to decide if it should be downloaded in bulk or one by one.
+    Checked images are split into two lists: small and large. Small images are downloaded in bulk, large images are downloaded one by one.
+    As soon as the task is created, it is put into the queue. Workers take tasks from the queue and execute them.
+
+    """
+    # to switch between single and bulk download
+    switch_size = kwargs.get("switch_size", 1.28 * 1024 * 1024)
```
Comment (almazgimaev) [inline supervisely/project/project.py:4426]: why 1.28 ? (where did we get this number?)

File: supervisely/api/image_api.py [hunk 4]
```diff
@@ -4157,3 +4165,185 @@ async def download_bytes_many_async(
...
+                    images = loop.run_until_complete(api.image.get_list_async(123456, per_page=600))
+        """
+
+        method = "images.list"
+        dataset_info = kwargs.get("dataset_info", None)
+
+        if dataset_info is None:
+            dataset_info = self._api.dataset.get_info_by_id(dataset_id, raise_error=True)
+
+        total_pages = ceil(dataset_info.items_count / per_page)
```
Comment (almazgimaev) [inline supervisely/api/image_api.py:4314]: ? mb need to use `first_response` to get items_count when using filters

File: supervisely/project/project.py [hunk 5]
```diff
@@ -4490,11 +4558,13 @@ async def _download_project_async(
                     only_image_tags=only_image_tags,
                     progress_cb=ds_progress,
                 )
-                tasks.append(task)
-            await asyncio.gather(*tasks)
+                await queue.put(task)
+
+        all_images = small_images + large_images
```
Comment (almazgimaev) [inline supervisely/project/project.py:4563]: is there a way to keep images order? (or it doesn't matter?)

File: supervisely/project/project.py (hunk 3 above)
Comment (GoldenAnpu) [reply supervisely/project/project.py:4426]: These are approximate recommended values for dividing files into large and small categories to fit within the recommended batch size. This scheme is a simple implementation of a more complex image dividing method. 
At the moment, this number is as it is, but we need to find an optimal value that will be effective in various scenarios.

File: supervisely/project/project.py (hunk 5 above)
Comment (GoldenAnpu) [reply supervisely/project/project.py:4563]: It doesn't matter for the following operations on the dataset. All images and annotations have already been saved before we start saving file metadata and deleting redundant items.

File: supervisely/api/module_api.py [hunk 6]
```diff
@@ -867,6 +869,96 @@ def _get_info_by_id(self, id, method, fields=None):
         response = self._get_response_by_id(id, method, id_field=ApiField.ID, fields=fields)
         return self._convert_json_info(response.json()) if (response is not None) else None
 
+    async def get_list_idx_page_async(
+        self,
+        method: str,
+        data: dict,
+    ) -> List[Any]:
```
Comment (almazgimaev) [inline supervisely/api/module_api.py:876]: typing

**Review by almazgimaev — APPROVED** [review]

**Pull request merged**